### PR TITLE
Moves DefinitionManagers to Core

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
This is a subset of PR #996. This includes only the changes needed to move the underlying DefinitionManagers/Builders into Core

## Related issues
Addresses [AB#73703](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/73703).

## Testing
All existing DefinitionManager unit tests must pass

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Features/Definition/CompartmentDefinitionManager.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            // The json file is a bundle compiled from the compartment definitions currently defined by HL7.
+            // The definitions are available at https://www.hl7.org/fhir/compartmentdefinition.html.
+            using Stream stream = _modelInfoProvider.OpenVersionedFileStream("compartment.json");
+            using TextReader reader = new StreamReader(stream);
+            using JsonReader jsonReader = new JsonTextReader(reader);
+            var bundle = new BundleWrapper(FhirJsonNode.Read(jsonReader).ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider));
+            Build(bundle);
+        }
+
+        public bool TryGetSearchParams(string resourceType, CompartmentType compartment, out HashSet<string> searchParams)
```
Comment (rbans96) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/CompartmentDefinitionManager.cs:64]: nit : compartment -> compartmentType

File: src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/BundleEntryWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        public BundleEntryWrapper(ITypedElement entry)
+        {
+            EnsureArg.IsNotNull(entry, nameof(entry));
+
+            _entry = entry;
+        }
+
+        public ITypedElement GetResource()
+        {
+            return _entry.Select("resource").FirstOrDefault();
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/BundleEntryWrapper.cs:26]: Would it be useful to cache this result? From our discussion it sounded like the Scalar/Select operations were high cost.

File: src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionBuilder.cs [hunk 3]
```diff
@@ -115,31 +116,34 @@ internal void Build()
...
+            return (resourceType == KnownResourceTypes.DomainResource && searchParameterName == "_text") ||
+                   (resourceType == KnownResourceTypes.Resource && searchParameterName == "_content") ||
+                   (resourceType == KnownResourceTypes.Resource && searchParameterName == "_query") ||
+                   ShouldExcludeEntryStu3(resourceType, searchParameterName);
+        }
+
+        private bool ShouldExcludeEntryStu3(string resourceType, string searchParameterName)
+        {
+            return _modelInfoProvider.Version == FhirSpecification.Stu3 &&
+                   (resourceType == "DataElement" &&
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionBuilder.cs:130]: Should DataElement also be a KnownResourceType? 
Nit: The parentheses on this line aren't needed.

File: src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+            _name = new Lazy<string>(() => searchParameter.Scalar("name")?.ToString());
+            _description = new Lazy<string>(() => searchParameter.Scalar("description")?.ToString());
+            _url = new Lazy<string>(() => searchParameter.Scalar("url")?.ToString());
+            _expression = new Lazy<string>(() => searchParameter.Scalar("expression")?.ToString());
+            _type = new Lazy<string>(() => searchParameter.Scalar("type")?.ToString());
+
+            _base = new Lazy<IReadOnlyList<string>>(() => searchParameter.Select("base")?.AsStringValues().ToArray());
+            _target = new Lazy<IReadOnlyList<string>>(() => searchParameter.Select("target")?.AsStringValues().ToArray());
+            _component = new Lazy<IReadOnlyList<ITypedElement>>(() => searchParameter.Select("component").ToArray());
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Definition/BundleWrappers/SearchParameterWrapper.cs:41]: Does this select need a null check like the other ones do?

**Pull request merged**